Sources/LLVM/Constant.swift

```diff
@@ -0,0 +1,490 @@
+  internal let llvm: LLVMValueRef
+  internal let repr: InternalConstantRepresentation
+  internal init(llvm: LLVMValueRef!) {
+    self.llvm = llvm
+
+    let reprID = ObjectIdentifier(Repr.self)
+    if reprID == ObjectIdentifier(Unsigned.self) {
+      self.repr = .unsigned
+    } else if reprID == ObjectIdentifier(Signed.self) {
+      self.repr = .signed
+    } else if reprID == ObjectIdentifier(Floating.self) {
+      self.repr = .floating
+    } else {
+      fatalError("Invalid representation \(type(of: Repr.self))")
+    }
+  }
+
+  /// Retrieves the underlying LLVM constant object.
+  public func asLLVM() -> LLVMValueRef {
+    return llvm
+  }
+
+  // MARK: Casting
+
+  /// Creates a constant cast to a given integral type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given
+  ///   integral type.
+  public func cast<T: IntegralConstantRepresentation>(to type: IntType) -> Constant<T> {
+    let destID = ObjectIdentifier(T.self)
+    let val = self.asLLVM()
+    if destID == ObjectIdentifier(Unsigned.self) {
+      switch self.repr {
+      case .unsigned: fallthrough
+      case .signed:
+        return Constant<T>(llvm: LLVMConstIntCast(val, type.asLLVM(), /*signed:*/ false.llvm))
+      case .floating:
+        return Constant<T>(llvm: LLVMConstFPToUI(val, type.asLLVM()))
+      }
+    } else if destID == ObjectIdentifier(Signed.self) {
+      switch self.repr {
+      case .unsigned: fallthrough
+      case .signed:
+        return Constant<T>(llvm: LLVMConstIntCast(val, type.asLLVM(), /*signed:*/ true.llvm))
+      case .floating:
+        return Constant<T>(llvm: LLVMConstFPToSI(val, type.asLLVM()))
+      }
+    } else {
+      fatalError("Invalid representation \(type(of: T.self))")
+    }
+  }
+
+  /// Creates a constant cast to a given floating type.
+  ///
+  /// - parameter type: The type to cast towards.
+  ///
+  /// - returns: A const value representing this value cast to the given 
+  ///   floating type.
+  public func cast(to type: FloatType) -> Constant<Floating> {
+    let val = self.asLLVM()
+    switch self.repr {
+    case .unsigned:
+      return Constant<Floating>(llvm: LLVMConstUIToFP(val, type.asLLVM()))
+    case .signed:
+      return Constant<Floating>(llvm: LLVMConstSIToFP(val, type.asLLVM()))
+    case .floating:
+      return Constant<Floating>(llvm: LLVMConstFPCast(val, type.asLLVM()))
+    }
+  }
+
+  // MARK: Arithmetic Operations
+
+  /// Creates a constant negate operation to negate a value.
+  ///
+  /// - parameter lhs: The operand to negate.
+  /// - parameter overflowBehavior: Should overflow occur, specifies the
+  ///   behavior of the resulting constant value.
+  ///
+  /// - returns: A constant value representing the negation of the given constant.
+  public static func neg(_ lhs: Constant<Signed>, overflowBehavior: OverflowBehavior = .default) -> Constant<Signed> {
```

Comment:
Sorry, back to my computer. I'm actually not sure why the spelled-out versions need to exist. But I think maybe they should be fully-spelled-out.